File: 3834. Merge Adjacent Equal Elements.cpp
class Solution {
public:
    vector<long long> mergeAdjacent(const vector<int>& nums) {
        stack<long long> st;
        for (int num : nums) {
            long long curr = num;
            while (!st.empty() && st.top() == curr) {
                curr += st.top();
                st.pop();
            }
            st.push(curr);
        }

        vector<long long> arr;
        while (!st.empty()) {
            arr.push_back(st.top());
            st.pop();
        }
        reverse(arr.begin(), arr.end());

        return arr;
    }
};

File: 3834. Merge Adjacent Equal Elements.java
class Solution {
    public List<Long> mergeAdjacent(int[] nums) {
        Stack<Long> st = new Stack<>();
        for(int num : nums){
            long curr = num;
            while(!st.isEmpty() && st.peek() == curr){
                curr += st.pop();
            }
            st.push(curr);
        }

        List<Long> arr = new ArrayList<>();
        while(!st.isEmpty()) arr.add(st.pop());
        Collections.reverse(arr);

        return arr;
    }
}